English Translation (#2)
alibaba/MNNKit: README.md

> ## 简介

Reviewer: ## Introduction

Reviewer: Add a link to the Chinese version of the README in the beginning

> MNNKit is a series of application-layer solution based off the device inference engine [MNN](https://github.com/alibaba/MNN), it's a mature solution after being massively deployed by the MNN Team in real world schenarious by the MNNKit. MNNKit mainly focus on Android/iOS Application and enabled them to rapidly and conviniently deploy AI capabilities to their applications so they can develop new apps on top of it.

Reviewer: ```suggestion
MNNKit is a collection of application-level solutions based on the on-device inference engine [MNN](https://github.com/alibaba/MNN). It's a mature solution after being massively deployed by the MNN team in real world scenarios. MNNKit is mainly for Android/iOS application developers and enables them to rapidly and conveniently deploy AI capabilities to their applications so they can develop new apps or features on top of it.
```

> - Usable out of the box. Easy to integrate and no prior knowledge of algorithm or model required.

Reviewer: ```suggestion
- Works out of the box. Easy to integrate without prior knowledge of algorithm or model.
```

open MNNKitDemo.xcworkspace
```

And install and run on a real device. (Video detection in the demo relies on Camera)

## SDK Installation

### Kit Dependencies

MNNKit SDK is organized in the following structure：

![SDK Stack](doc/sdk_stack.png)

> Down-top divided into three layers:

Reviewer: ```suggestion
Divided into three layers from the bottom to the top:
```

> 1. MNN Enginer Layer is a prebuilt Release version of the [MNN](https://github.com/alibaba/MNN) open-source library on Android/iOS which provides the execution environment of on-device AI.

Reviewer: ```suggestion
1. MNN Enginer Layer is a prebuilt release version of the [MNN](https://github.com/alibaba/MNN) open-source library on Android/iOS which provides the execution environment of on-device AI.
```

> 2. Core layer is layer that provides platform native binding of the underlying MNN library. This layer is a wrapper around of the C++ APIs in Objective-C (iOS) or Java (Android. WIP) as well as providing some public Utility class and definitions for the upper-level SDK.

Reviewer: ```suggestion
2. Core layer is layer that provides platform native binding of the underlying MNN library. This layer is a wrapper around of the C++ APIs in Objective-C (iOS) or Java (Android) (both APIs are WIP) as well as providing some public utility class and definitions for the upper-level SDK.
```

> 3. Kit layer. FaceRecognition/HandGestureDetection etc are all abstract wrappers of a specific algorithm. It contains multiple models and their required processing. This layer will be expanded with more models and capabilities as time goes on.

Reviewer: ```suggestion
3. Kit layer. FaceRecognition/HandGestureDetection etc are all abstract wrappers of specific algorithms. It contains multiple models and their respective processing. This layer will be expanded with more models and capabilities as time goes on.
```

### Installation

> SDKs at the kit layer are independnet of each other and has implicit dependencies to the required components below it and doesn't explicit specification by the end-user.

Reviewer: ```suggestion
SDKs at the kit layer are independnet of each other and has implicit dependencies to the required components below it and doesn't need explicit specifications by the end-user.
```

#### Android

> - Required minimum API Level16（Android 4.1）

Reviewer: ```suggestion
- Required minimum API Level 16（Android 4.1）
```

#### iOS

- Required minimum version iOS 8.0

```ruby
source 'https://github.com/CocoaPods/Specs.git'
platform :ios

target 'MNNKitDemo' do
    platform :ios, '8.0'

    pod 'MNNFaceDetection', '~> 0.0.1'
    pod 'MNNHandGestureDetection', '~> 0.0.1'
    pod 'MNNPortraitSegmentation', '~> 0.0.1'

end
```

##### Bitcode

MNNKit SDKs doesn't support Bitcode and you'll need to turn off Bitcode in your application.

## API

There are only three basic APIs in the kit:
  - Create Instance
  - Inference
  - Release Instance
And this is also the sequence for using it, as shown in the graphs below. You can use video/image or other formats of data as the input of the inferencing stage.

![api](doc/api_flow.png)

[FaceDetection API](doc/FaceDetection_CN.md)

[HandGestureDetection API](doc/HandGestureDetection_CN.md)

[PortraitSegmentation API](doc/PortraitSegmentation_CN.md)

> 

Reviewer: We haven't translated these yet, have we?

## Integration Guide

> MNNKit API contains two arguments: ``inAngle`` and ``outAngle``. What are they used for? Because understanding this question, we should first understanding the generic process of how the SDK works.

Reviewer: ```suggestion
MNNKit API contains two arguments: ``inAngle`` and ``outAngle``. What are they used for? To answer this question, we should first understand the general process of how the SDK works.
```

> ### Overview

Reviewer: ```suggestion
### Sequence of Operations
```

> Below is the whole process of processing what the back camera of iOS and Android devices has captured.

Reviewer: ```suggestion
Below is the sequence to process what the back-facing camera on iOS and Android devices has captured.
```

![process](doc/process_graph.png)

#### 1. Device preview and output

This stage captures data from the camera and feed it into the SDK while providing proper preview on the screen.

> #### 2. SDK Detection

Reviewer: ```suggestion
#### 2. Detection in the SDK
```

> MNNKit SDK contains the deep learning model of related algorithms and runs on top of MNN engine. As a result it's not traditional image processing and contains all the characteristics of deep learning. During training stage algorithm uses face images in top-down direction as such this model only recognize faces in that same direction. If the input contains faces in a different direction, it might not get detected. We call this ``Content Orientation Sensitive``. In MNNKit, FaceDetection,HandGestureDetection and PortraitSegmentation are all ``Content Orientation Sensitive``.

Reviewer: ```suggestion
MNNKit SDK contains the deep learning models of related algorithms and runs on top of MNN engine. As a result it's not traditional image processing and contains all the characteristics of deep learning. During the training stage, the algorithm uses face images in upright orientation. As a result, this model only recognizes faces in that orientation. If the input contains faces in a different orientation, it might not get detected. We call this ``Content Orientation Sensitive``. In MNNKit, FaceDetection,HandGestureDetection and PortraitSegmentation are all ``Content Orientation Sensitive``.
```

> 1. Prior to inferencing with MNN, pre-processing is needed on the raw input to make sure that faces in the raw input image has the right direction.

Reviewer: ```suggestion
1. Prior to inferencing with MNN, pre-processing is needed on the raw input to make sure that faces in the raw input image has the correct orientation.
```

2. The inference process yields results based on the coordinate system of the pre-processed image.
3. To provide easy-to-use results for the caller, the SDK transforms the coordinate system to the same as the screen rendering coordinate system.

> ``inAngle`` in the SDK arguments is to control the angle to rotate so that the raw input in rotated to the required direction and ``outAngle`` is the angle used to rotate the raw inferenced coordinates to the screen rendering coordinate system.

> As we can see in the image, for a device in positive direction, the back camera has ``inAngle`` and ``outAngle`` both set to ``0`` for iOS, and for Android it's ``90`` and ``0`` respectively.

Reviewer: ```suggestion
As we can see in the image, if you take a picture with the back camera in portait mode, the image has ``inAngle`` and ``outAngle`` both set to ``0`` for iOS, and for Android it's ``90`` and ``0`` respectively.
```

#### 3. Result rendering

> The final result of the SDK is the description of the feature points, or the coordinate of the feature points to be precise. All these coordinates are related to a certain coordinate system, a system that uses the left-top point of image as the origin. For example if the input image is of resolution ``1280*720``, then the result coordinates are also based on this exact coordinate system.

Reviewer: ```suggestion
The final result of the SDK is the description of the feature points, or the coordinate of the feature points to put it simply. All these coordinates are related to the image coordinate system, which uses the top-left point of image as the origin. For example, if the input image is of resolution ``1280*720``, then the result coordinates are also based on this exact coordinate system.
```

> In real world applications, the resulting feature points are supposed to be displayed on the user's screen. The frontend would uses a canvas for rendering which could be a view in the UI that is used for displaying or it could be some other components. The size of the canvas is specified by the application itself based on UI design. The coordinate system of this canvas is called rendering canvas. What this step is doing is to convert the raw inferenced points to rendering coordinate system.

Reviewer: ```suggestion
In real world applications, the resulting feature points are supposed to be displayed on the user's screen. The frontend would use a canvas for rendering which could be a view in the UI that is used for displaying or it could be some other components. The size of the canvas is specified by the application itself based on UI design. The coordinate system of this canvas is called rendering coordinates. What this step is doing is to convert the feature points from the image's coordinate system to rendering coordinate system.
```

> In the last step of the SDK, we've already transformed the feature points into the same direction as the rendering coordinate system. Since the size of the two coordinate system is different, all that's left to do is to scale the coordinates accordingly followed by rendering them to the canvas and finish the whole process.

Reviewer: ```suggestion
In the last step of the SDK's detection process, we've already transformed the feature points into the same orientation as the rendering coordinate system. Since the size of the two coordinate system is different, all that's left to do is to scale the coordinates accordingly followed by rendering them to the canvas and finish the whole process.
```

> ### Engineering practice reference

Reviewer: ```suggestion
### Best Practice
```

> In the aforementioned example we used the back camera in the top-down direction as an example and is relatively easier. However in real world applications,camera angle,device rotation and some other elements might change the direction of the image and need to be resolved outside the SDK.

Reviewer: ```suggestion
In the aforementioned example, we used the back camera in the upright orientation as an example and is relatively easy. However in real world applications, camera angle, device rotation and some other elements might change the direction of the image and need to be resolved outside the SDK.
```

> MNNKit Demo contains a generic solution to many of those problems and should be as a reference.

Reviewer: ```suggestion
MNNKit Demo contains solutions to many of those problems in practice, such as input image angle, camera preview, result rendering, auto-rotation and should be as a best practice guide. If you are unfamiliar with how to deal with these problems, please read coding samples in the demo. We believe that'll solve all your problems.
```

MNNKit Demo中涵盖了工程实践中所有问题的综合解法，如输入角度、摄像头预览、结果渲染、自动旋转等等，是工程接入的最佳实践参考。如果不熟悉该如何处理，请阅读Demo中的代码示例，相信所有的问题都可以迎刃而解。
